File: java/0438-find-all-anagrams-in-a-string/Solution2.java
import java.util.ArrayList;
import java.util.List;

public class Solution2 {
    public List<Integer> findAnagrams(String s, String p) {
        int pLength = p.length();
        int sLenght = s.length();
        if (sLenght < pLength) {
            return new ArrayList<>();
        }

        int[] counts = new int[26];
        for (int i = 0; i < pLength; ++i) {
            ++counts[p.charAt(i) - 'a'];
        }

        List<Integer> result = new ArrayList<>();

        int required = pLength;
        for (int left = 0, right = 0; right < sLenght; ++right) {
            if (--counts[s.charAt(right) - 'a'] >= 0) {
                --required;
            }

            while (required == 0) {
                if (right - left + 1 == pLength) {
                    result.add(left);
                }

                if (++counts[s.charAt(left++) - 'a'] > 0) {
                    ++required;
                }
            }
        }

        return result;
    }
}

File: c++/0438-find-all-anagrams-in-a-string/solution2.cpp
#include <string>
#include <vector>

using namespace std;

class Solution2 {
public:
    vector<int> findAnagrams(string s, string p) {
        int pSize = p.size();
        int sSize = s.size();
        if (sSize < pSize) {
            return {};
        }

        int counts[26] = {0};
        for (int i = 0; i < pSize; ++i) {
            ++counts[p[i] - 'a'];
        }

        vector<int> result;

        int required = pSize;
        for (int left = 0, right = 0; right < sSize; ++right) {
            if (--counts[s[right] - 'a'] >= 0) {
                --required;
            }

            while (required == 0) {
                if (right - left + 1 == pSize) {
                    result.push_back(left);
                }

                if (++counts[s[left++] - 'a'] > 0) {
                    ++required;
                }
            }
        }

        return result;
    }
};
